sae/vm_test.go

```diff
@@ -515,3 +537,195 @@ func TestSemanticBlockChecks(t *testing.T) {
+			go func() {
+				tb.Helper()
+
+				assert.Contains(tb, peers, peerID, "unknown peer in SendAppResponse")
+				peer, ok := peers[peerID]
+				if !ok {
+					return
+				}
+
+				assert.NoErrorf(tb, peer.AppResponse(ctx, nodeID, r, b), "sending response to %s", peerID)
+			}()
+
+			return nil
+		}
+		sender.SendAppErrorF = func(ctx context.Context, peerID ids.NodeID, r uint32, code int32, msg string) error {
+			go func() {
+				tb.Helper()
+				assert.Contains(tb, peers, peerID, "unknown peer in SendAppError")
+				peer, ok := peers[peerID]
+				if !ok {
+					return
+				}
+
+				appErr := &snowcommon.AppError{
+					Code:    code,
+					Message: msg,
+				}
+				assert.NoErrorf(tb, peer.AppRequestFailed(ctx, nodeID, r, appErr), "sending error to %s", peerID)
+			}()
+			return nil
+		}
+		sender.SendAppGossipF = func(ctx context.Context, c snowcommon.SendConfig, b []byte) error {
+			go func() {
+				tb.Helper()
+
+				for peerID := range c.NodeIDs {
+					if peer, ok := peers[peerID]; ok {
+						assert.NoErrorf(tb, peer.AppGossip(ctx, nodeID, b), "sending gossip to %s", peerID)
+					}
+				}
+				var sent set.Set[ids.NodeID]
+				sent.Union(c.NodeIDs)
+
+				send := func(peers map[ids.NodeID]*SUT, count int) error {
+					for peerID, peer := range peers {
+						if count <= 0 {
+							break
+						}
+						if sent.Contains(peerID) {
+							continue
+						}
+						if err := peer.AppGossip(ctx, nodeID, b); err != nil {
+							return err
+						}
+
+						sent.Add(peerID)
+						count--
+					}
+					return nil
+				}
+
+				assert.NoError(tb, send(validatorNodes, c.Validators), "sending to validators")
+				assert.NoError(tb, send(nonValidatorPeers, c.NonValidators), "sending to non-validators")
+				assert.NoError(tb, send(peers, c.Peers), "sending to peers")
+			}()
+			return nil
+		}
+
+		// Connect all the peers _after_ setting up the sender functions.
+		defer func() {
+			tb.Helper()
+			for peerID := range peers {
+				require.NoErrorf(tb, sut.Connected(tb.Context(), peerID, version.Current), "Connected(%s)", peerID)
+			}
+		}()
+	}
+
+	return &networkSUT{
+		validators:    slices.Collect(maps.Values(validatorNodes)),
+		nonValidators: slices.Collect(maps.Values(nonValidatorNodes)),
+	}
+}
+
+func requireReceiveTx(tb testing.TB, nodes []*SUT, txHash common.Hash) {
+	tb.Helper()
+	for _, sut := range nodes {
+		assert.Eventuallyf(tb, func() bool {
+			return sut.rawVM.mempool.Has(ids.ID(txHash))
+		}, 5*time.Second, 100*time.Millisecond, "tx %x not gossiped to node %s", txHash, sut.rawVM.snowCtx.NodeID)
+	}
+	require.False(tb, tb.Failed())
+}
+
+func requireNotReceiveTx(tb testing.TB, nodes []*SUT, txHash common.Hash) {
+	tb.Helper()
+	for _, sut := range nodes {
+		assert.False(tb, sut.rawVM.mempool.Has(ids.ID(txHash)), "tx %x was gossiped to node %s", txHash, sut.rawVM.snowCtx.NodeID)
+	}
+	require.False(tb, tb.Failed())
+}
+
+func TestGossip(t *testing.T) {
```

Comment:
It feels a bit crazy that we have like 200 lines of setup for such a small test, but hopefully we'll extend the usage of the `networkSUT` (or replace it with Josh's setup)